source/common/router/router.cc

```diff
@@ -492,6 +494,33 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
 
   transport_socket_options_ = Network::TransportSocketOptionsUtility::fromFilterState(
       *callbacks_->streamInfo().filterState());
+
+  auto has_downstream_connection_with_redirect_records =
```

Comment:
One idea (although I'm not sure if it pans out): could `buildWFPRedirectRecordsOptions` be called when this filterState() is set, and instead of putting in a `Win32RedirectRecordsFilterState`, put in a generic list of socket options that should be applied to the upstream? `Network::UpstreamSocketOptionsFilterState` or something. It would reduce a bit of duplication between router and tcp_proxy, and would encapsulate more of the original_dst/win32 details in the original_dst filter.

Reply:
fixed in https://github.com/envoyproxy/envoy/pull/14753/commits/1e5b32387ef971a8d84363bfe7ac93f01270b0db